Rev3 mainboard: Add initial support for the KSZ9896 & mainboard detection

AndreRenaud opened the pull request:
Preliminary testing on rev2 & rev3 mainboards has been done

AndreRenaud committed:
Rev3 mainboard: Add initial support for the KSZ9896 & mainboard detection

AndreRenaud committed:
FDT fixup - use slightly better mechanism

AndreRenaud committed:
KSZ9896: Fix up single-led mode and MMD access

LED behaviour now matches rev2 Marvell behaviour

AndreRenaud committed:
Enable checksum commands

sjg20 commented on board/designa/snappermx6/ksz9896.c:
```diff
@@ -0,0 +1,215 @@
+ #define KSZ9896_MMD_LED_MODE                                   0x02, 0x00
+ #define KSZ9896_MMD_EEE_ADV                                    0x07, 0x3C
+
+static int ksz9896_write16(struct udevice *dev, uint8_t port, uint16_t reg, uint16_t val)
```
This driver should probably be in drivers/net/phy

AndreRenaud replied:
Yes, that is true. I was waiting to see what the Linux .dts file ended up looking like and then replicating it in U-Boot, but I should have started with it there.

sjg20 replied:
OK

sjg20 commented on board/designa/snappermx6/ksz9896.c:
```diff
@@ -0,0 +1,215 @@
+	uint32_t id = 0;
+    uint8_t xmii_ctrl_1 = 0x18;
+
+	if (i2c_get_chip_for_busnum(0, 0x5f, 2, &dev) < 0)
```
You should be able to use the phy subsystem

Compatible string is marvell,88E1510

U_BOOT_PHY_DRIVER() to declare it

I'm not sure that the phy subsystem does driver model very well though...

sjg20 commented on board/designa/snappermx6/ksz9896.c:
```diff
@@ -0,0 +1,215 @@
+		return 1;
+	}
+
+
```
try patman to check the patches

sjg20 commented on board/designa/snappermx6/snappermx6.c:
```diff
@@ -39,60 +40,75 @@ static iomux_v3_cfg_t const uart5_pads[] = {
+	// tree being initialised, so the dm_ functions will not work
+	// We're on GPIO bank 2, pins 9 & 10
+
+	uint32_t gpio_state = readl(GPIO2_BASE_ADDR);
```
This should use the GPIO driver...also there is 

gpio_request_list_by_name()
dm_gpio_get_values_as_int()

AndreRenaud replied:
Does this work for things that need to run prior to device tree set up? The trick here is that we're using this pin to modify the device tree, so it needs to be done very early?

sjg20 replied:
Do you have an SPL? If so, it could perhaps be done there?

If not, could it be done in pre-relocation code, so that DM is setup. Then you would have the right fdt for post-relocation

AndreRenaud replied:
Yes, we're using the U-Boot SPL

sjg20 commented on board/designa/snappermx6/snappermx6.c:
```diff
@@ -39,60 +40,75 @@ static iomux_v3_cfg_t const uart5_pads[] = {
-	SETUP_IOMUX_PADS(enet_pads);
-
-	/* Reset the 88e6061 PHY on the Salmon carrier board */
+	int version = salmon_version_read();
```
This should probably be a sysinfo driver

AndreRenaud replied:
I'm not familiar with this. Do you mean like `omap3_sysinfo`, using `checkboard` etc...? That doesn't seem to be a generic piece of infrastructure?

sjg20 replied:
It's in sysinfo.h - basically a driver that returns system info

AndreRenaud replied:
sysinfo.h appears to be x86/arch-coreboot specific?

AndreRenaud committed:
ksz9896: Migrate to drivers/net/phy & bind in via device tree

AndreRenaud committed:
Reinstate i2c address & fix KSZ error messages

AndreRenaud committed:
Salmon rev3b: Updated ethernet reset polarity

AndreRenaud committed:
workflow: Move to latest versions of the actions

AndreRenaud committed:
SnapperMX6: Added multi-fit image support via environment variable

AndreRenaud committed:
Salmon: Use FIT_BEST_MATCH to select kernel device tree automatically

Should make it easier to maintain compatibility with older releases, and
keep the previous bootcmd functional

AndreRenaud committed:
salmon: reinstated support for older boot images with previous compatibility strings
